crates/metadata-providers/src/dynamodb.rs

```diff
@@ -0,0 +1,492 @@
+// Copyright (c) 2023 - 2025 Restate Software, Inc., Restate GmbH.
+// All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use aws_config::{BehaviorVersion, Region, timeout::TimeoutConfig};
+use aws_sdk_dynamodb::{
+    config::Credentials,
+    error::SdkError,
+    operation::{delete_item::DeleteItemError, put_item::PutItemError},
+    primitives::Blob,
+    types::{AttributeValue, ReturnConsumedCapacity, ReturnValue},
+};
+use bytestring::ByteString;
+
+use restate_metadata_store::{ProvisionedMetadataStore, ReadError, WriteError};
+use restate_types::{
+    Version,
+    config::{DynamoDbOptions, MetadataClientOptions},
+    metadata::{Precondition, VersionedValue},
+    net::connect_opts::CommonClientConnectionOptions,
+};
+
+const KIND_FIELD: &str = "kind";
+const VERSION_FIELD: &str = "version";
+const DATA_FIELD: &str = "data";
+
+const CONDITION_DOES_NOT_EXIST: &str =
+    const_format::concatcp!("attribute_not_exists(", KIND_FIELD, ")");
+
+#[derive(Debug, thiserror::Error)]
+pub enum DynamoDbStoreCodecError {
+    #[error("Missing field {0}")]
+    MissingField(&'static str),
+    #[error("Field {0} has wrong type")]
+    InvalidType(&'static str),
+    #[error("Field {0} has an invalid format")]
+    InvalidFormat(&'static str),
+}
+
+impl From<DynamoDbStoreCodecError> for ReadError {
+    fn from(value: DynamoDbStoreCodecError) -> Self {
+        Self::Codec(value.into())
+    }
+}
+pub struct DynamoDbMetadataStore {
+    table: String,
+    key_prefix: Option<String>,
+    client: aws_sdk_dynamodb::Client,
+}
+
+impl DynamoDbMetadataStore {
+    pub async fn new(
+        table: impl Into<String>,
+        key_prefix: Option<String>,
+        config: &MetadataClientOptions,
+        dynamo_db_options: DynamoDbOptions,
+    ) -> anyhow::Result<Self> {
+        let timeout_config = TimeoutConfig::builder()
+            .connect_timeout(config.connect_timeout())
+            .build();
+
+        // todo: figure out proper configuration.
+        let cfg = aws_config::defaults(BehaviorVersion::latest()).timeout_config(timeout_config);
```

Comment:
What's the default timeout config? At other places in the code base we used the default.

Reply:
I mainly wanted to respect all configuration in `MetadataClientOptions` as much as possible. Not all config applies to dynamo-db (for example keep-alive-timeout).

But after reviewing this it might be better to just use the config default to avoid disabling other timeouts (read and operation timeout)